libraries/chain/controller.cpp

```diff
@@ -520,6 +522,28 @@ struct controller_impl {
          const auto hash = calculate_integrity_hash();
          ilog( "database initialized with hash: ${hash}", ("hash", hash) );
       }
+
+      if (head && fork_db.pending_head() && fork_db.root()) {
+         if (read_mode != db_read_mode::IRREVERSIBLE) {
+            if (head->block_num < fork_db.pending_head()->block_num) { 
+               // irreversible mode => speculative mode
```

Comment:
On second thought, trying to keep the reversible blocks database in sync with the best fork_db branch is unnecessary, adds extra code, and may even hurt sync performance, all for essentially no gain. I still think that `maybe_switch_forks` should not be called during irreversible mode, so I now suggest the following.

In irreversible mode, the reverse block database should become and remain completely empty (you could just purge it if necessary at statup). Only the fork_db.dat is used as the source for any reversible blocks. If necessary, undo chainbase db state enough to reach LIB which is where irreversible mode should always start.

When switching from irreversible mode to speculative mode, it will be necessary to populate the reversible blocks database with the best branch from fork_database, which is to say the branch that goes from the root to the pending_head. This should be done before the part of the code that calls replay, which will then apply regular replay logic to have it catch up from LIB to head. No need for maybe_switch_forks.